### PR TITLE
[Cherrypick 0.13] Add M1 build and test jobs

### PR DESCRIPTION
**Pull request by YosuaMichael**
Cherrypick from main: #5948 and #6110

This change will add a file `.github/workflows/build-m1-binaries.yml` which build m1 on the CI.

EDIT: this more generally picks all M1-related stuff, including https://github.com/pytorch/vision/pull/6122  and https://github.com/pytorch/vision/pull/6132 (and a lot more stuff tracked in https://github.com/pytorch/vision/issues/6081#issuecomment-1141874556)

### CODE REVIEW

**Review by NicolasHug — APPROVED** [review]
Thanks Yosua

**Comment by YosuaMichael** [comment]
> Thanks Yosua
Fast! Thanks!

**Comment by datumbox** [comment]
FYI we will need to also cherrypick https://github.com/pytorch/vision/pull/6111

**Comment by YosuaMichael** [comment]
> #6111

@datumbox once #6111 is merged on main, we can put it together on this PR

**Review by malfet — CHANGES_REQUESTED** [review]
It needs some more changes to be usable on release branches, please hold until they are done. (For example, one need to change version from 0.14 back to 0.13)

**Review by datumbox — DISMISSED** [review]
@NicolasHug grabbed the latest differences from numerous M1 related PRs merged on main.

I understand that this is a straight copy-paste of the final configs from main. They all LGTM.

**Comment by NicolasHug** [comment]
Before merging we will need @malfet 's input on https://github.com/pytorch/vision/pull/6117#issuecomment-1147400892.

https://github.com/pytorch/vision/pull/6117 has made changes to `.github/workflows/build-m1-binaries.yml` and I wonder if the env variables are correctly set now.

**Comment by datumbox** [comment]
@NicolasHug I think his PR at #6117 does solve the issue and if it's not already cherrypicked, needs to be included here. I agree, let's wait to see what he thinks.

File: .github/workflows/build-m1-binaries.yml
Comment (NicolasHug) [inline .github/workflows/build-m1-binaries.yml:35]: I have edited this line to rely on the test repo instead of the nightlies, but there are still various references to "nightly" in this script and I'm not sure what to do with them.

@malfet would you mind double checking this whole file?
(perhaps you'd like to wait after https://github.com/pytorch/vision/pull/6140?)

File: .github/workflows/build-m1-binaries.yml
Comment (NicolasHug) [inline .github/workflows/build-m1-binaries.yml:33]: @malfet @atalman 

It looks like the builds are using the nightly channel instead of the test channel, see e.g. https://github.com/pytorch/vision/runs/6858724488?check_suite_focus=true#step:4:133

Should we update the tag to also consider branches like the current `release/0.13`?

File: .github/workflows/build-m1-binaries.yml
Comment (datumbox) [inline .github/workflows/build-m1-binaries.yml:82]:
```suggestion
        if: ${{ github.event_name == 'push' && (github.event.ref == 'refs/heads/nightly' || startsWith(github.event.ref, 'refs/tags/')) }}
```

File: .github/workflows/build-m1-binaries.yml
Comment (datumbox) [inline .github/workflows/build-m1-binaries.yml:142]:
```suggestion
        if: ${{ github.event_name == 'push' && (github.event.ref == 'refs/heads/nightly' || startsWith(github.event.ref, 'refs/tags/')) }}
```

File: .github/workflows/build-m1-binaries.yml
Comment (malfet) [reply .github/workflows/build-m1-binaries.yml:82]: This is irrelevant for release branch, as we are not going to build nightlies from release branch anyway...

**Comment by NicolasHug** [comment]
> At this point, with so many changes, it might be worth checking that this PR contains all possible improvements done on main related to M1s.

I tried to keep track of everything and updated https://github.com/pytorch/vision/issues/6081#issuecomment-1141874556, but yes. There are so many PRs that this will require a very careful review.

**Comment by datumbox** [comment]
@NicolasHug Agreed. Happy to lend a pair of eyes at the last PR if you want. Ping me when ready.

**Review by NicolasHug — CHANGES_REQUESTED** [review]
The build jobs are currently failing with a dependency resolution conflict https://github.com/pytorch/vision/runs/6864443482?check_suite_focus=true

Marking red so we don't merge by mistake.

File: .github/workflows/build-m1-binaries.yml
Comment (atalman) [reply .github/workflows/build-m1-binaries.yml:34]: Done merged fix for this

File: .github/workflows/build-m1-binaries.yml
Comment (atalman) [reply .github/workflows/build-m1-binaries.yml:110]: Done! Fix is commited: https://github.com/pytorch/vision/pull/6112/commits/fb9a761abac621f16b50ebb01204fc31fae1d3bd

**Comment by datumbox** [comment]
@NicolasHug @YosuaMichael  There is another PR that adds M1 support by @atalman at #6167. Might be worth merging that one instead.

**Pull request closed without merging**
